app/Http/Controllers/IcsController.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Models\IcsToken;
+use Carbon\Carbon;
+use Illuminate\Http\RedirectResponse;
+use Illuminate\Http\Request;
+use Illuminate\Http\Response;
+use Illuminate\Support\Str;
+use Spatie\IcalendarGenerator\Components\Calendar;
+use Spatie\IcalendarGenerator\Components\Event;
+
+class IcsController extends Controller
+{
+    public function renderIcs(Request $request): Response {
+        $validated = $request->validate([
+                                            'user_id' => ['required', 'exists:users,id'],
+                                            'token'   => ['required', 'exists:ics_tokens,token'],
+                                            'limit'   => ['nullable', 'numeric', 'gte:1', 'lte:10000'],
+                                            'from'    => ['nullable', 'date'],
+                                            'to'      => ['nullable', 'date']
+                                        ]);
+
+        $icsToken = IcsToken::where('token', $validated['token'])->first();
```

Comment:
could this be done with a `firstOrFail()`?

Reply:
This is not needed because it is already checked in the validator.

Reply:
So then the first comparison in Line 27 is kinda redundant, isn't it?

Reply:
The `== null` part is redundant, yes. But the check against the user_id is needed.

Reply:
Okay. I'll approve this PR but please remove the `$icsToken == null ||`. You can then merge it yourself. :)